walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,58 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    List<TimeRange> optionalEventTimes = new ArrayList<>();
+    List<TimeRange> eventTimes = new ArrayList<>();
+    events.stream()
+          .sorted((e1, e2) -> e1.getWhen().start() - e2.getWhen().start())
+          .forEach(e -> {
+            boolean attendee = !Collections.disjoint(e.getAttendees(), attendees);
+            boolean optional = !Collections.disjoint(e.getAttendees(), optionalAttendees);
+            if (optional || attendee) {
+              optionalEventTimes.add(e.getWhen());
+            } 
+            if (attendee) {
+              eventTimes.add(e.getWhen());
+            }
+          });
+    
+    List<TimeRange> optionalTimes = getOpenTimes(optionalEventTimes, request.getDuration());
+    List<TimeRange> mandatoryTimes = getOpenTimes(eventTimes, request.getDuration());
+    return optionalTimes.size() == 0 && attendees.size() != 0 ? mandatoryTimes : optionalTimes;
+  }
+
+  private List<TimeRange> getOpenTimes(List<TimeRange> eventTimes, long duration) {
+    // Add value to end of list to allow safe "peeking" at the next element in loop.
+    eventTimes.add(TimeRange.fromStartDuration(TimeRange.END_OF_DAY, 0));
```

Comment:
It would be useful to add a parameter comment here, as the value of 0 is unclear. 
`eventTimes.add(TimeRange.fromStartDuration(TimeRange.END_OF_DAY, /* duration= */ 0));`

Reply:
Thank you! I was unaware and will include these in my code from now on. Fix pushed.